src/tactic/core.lean

```diff
@@ -1156,17 +1172,22 @@ adds an instance `ring new_int`, defined to be the instance of `ring ℤ` found
 Multiple instances can be added with `@[derive [ring, module ℝ]]`.
 -/
 @[derive_handler] meta def delta_instance : derive_handler :=
-λ cls tp,
-(do tp' ← mk_const tp,
-   tgt ← to_expr ``(%%cls %%tp'),
-   (_, v) ← solve_aux tgt (delta_target [tp] >> apply_instance >> done),
-   v ← instantiate_mvars v,
-   nm ← get_unused_name $ tp ++
-     match tgt with
-     | expr.app (expr.const nm _) _ := nm
+λ cls new_decl_name,
+(do new_decl_type ← declaration.type <$> get_decl new_decl_name,
+   new_decl_pexpr ← resolve_name new_decl_name,
+   tgt ← to_expr $ apply_under_pis cls new_decl_pexpr new_decl_type,
+   (_, inst) ← solve_aux tgt
+     (intros >> reset_instance_cache >> delta_target [new_decl_name]  >> apply_instance >> done),
+   inst ← instantiate_mvars inst,
+   tgt ← instantiate_mvars tgt,
+   nm ← get_unused_name $ new_decl_name ++
```

Comment:
A reminder to myself to change this when #1402 lands (https://github.com/leanprover-community/mathlib/pull/1402#issuecomment-535174226).